{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66ddebc8-b861-4ab0-bcb9-ef0c64b34a0a",
   "metadata": {},
   "source": [
    "# 1. Preprocess Data\n",
    "\n",
    "In this notebook we preprocess genomic data associated with R-loops originally generated by the [Arsuaga–Vázquez Lab](https://arsuaga-vazquez-lab.faculty.ucdavis.edu) at UC Davis and reported in the paper *“The R-loop grammar predicts R-loop formation under different topological constraints”* [by Ferrari *et al.*](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1013376) The raw data used here can be accessed from the [Zenodo repository](https://zenodo.org/records/15742754) as the file `fasta_bed.zip`.\n",
    "\n",
    "The experimental data was obtained using **single-molecule R-loop footprinting and sequencing (SMRF-seq)**. This technique identifies the genomic start and end positions of R-loops formed on DNA plasmids under different initial topologies.\n",
    "\n",
    "Each experiment is defined by a combination of the following parameters:\n",
    "\n",
    "- **Plasmid**:  \n",
    "  Two different DNA plasmids are used: **pFC53** and **pFC8**.\n",
    "\n",
    "- **DNA topology**:  \n",
    "  Plasmids are prepared in different topological states prior to transcription:\n",
    "  - **Linearized**: Circular plasmids are linearized before transcription.\n",
    "  - **Supercoiled**: Plasmids retain their native bacterial supercoiling, with supercoiling density approximately $\\sim -0.07$.\n",
    "  - **Gyrase-treated (Gyrasecr)**: Plasmids are treated with gyrase before transcription, increasing the negative supercoiling density to approximately $\\sim -0.14$.\n",
    "\n",
    "- **R-loop position**:  \n",
    "  For each molecule, SMRF-seq identifies the **starting** and **ending** coordinates of the R-loop along the plasmid sequence. In this notebook, we focus on the **starting locations** of R-loops.\n",
    "\n",
    "\n",
    "We use this data to construct libraries of short DNA sequences centered at the starting positions of R-loops. In subsequent notebooks, we show that these collections of words are structurally distinct: topological signatures derived from their **Insertion Chain Complexes** can differentiate between sets of words associated with R-loops formed under linear versus supercoiled DNA topologies.\n",
    "\n",
    "### Construction of sequence libraries\n",
    "\n",
    "We build these libraries as follows:\n",
    "\n",
    "1. For each plasmid and each R-loop starting site, we extract a 14-mer window of DNA sequence centered at the starting position of the R-loop. That is, if $a_1$ denotes the first nucleotide of the R-loop, we consider the window $$a_{-7}a_{-6}\\dots a_{-1}a_1a_2\\dots a_6a_7,$$  centered at the R-loop start.\n",
    "\n",
    "2. From this window, we extract all contiguous $k$-mers that include the start position, for $k=1, \\dots, 14$:\n",
    "\n",
    "   - **Length 1**:  $a_{-1}, \\quad a_1$\n",
    "\n",
    "   - **Length 2**:  $a_{-2}a_{-1}, \\quad a_{-1}a_1, \\quad a_1a_2$\n",
    "\n",
    "   - **Length 3**:  $a_{-3}a_{-2}a_{-1},\\quad a_{-2}a_{-1}a_1,\\quad a_{-1}a_1a_2, \\quad a_1a_2a_3$\n",
    "\n",
    "   - $\\vdots$\n",
    "\n",
    "   - **Length 14**:  $a_{-7}a_{-6}\\dots a_{-1}a_1\\dots a_6a_7$\n",
    "  \n",
    "3. We aggregate all extracted $k$-mers across all R-loop start sites for a given experimental condition and compute their frequencies.\n",
    "\n",
    "4. We normalize these frequencies by dividing each value by the maximum observed frequency, obtaining weights in the interval $(0,1]$.\n",
    "\n",
    "5. We retain only the $N = 500$ words with the highest normalized frequencies to form the representative sequence library for that condition.\n",
    "\n",
    "The resulting processed datasets are saved as `.pickle` files in the `data/preprocessed` directory for efficient use in the subsequent steps of the analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "95a23553-2621-49bc-83ba-2fb6c80c9613",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make sure biopython is installed\n",
    "#!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "41465039-5790-410d-aed8-d1182801f6d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.preprocess import read_subsequence, produce_table, get_words_combined\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b5d75d8-d5be-4b55-a7f2-43ef8ef62bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder paths in the notebook\n",
    "folder_path_bed = \"data/raw/fasta_bed/bed-files/\"\n",
    "folder_path_fasta = \"data/raw/fasta_bed/fasta-files/\"\n",
    "\n",
    "max_N_words=500\n",
    "max_window_side=7\n",
    "\n",
    "# FASTA mapping\n",
    "dictionary_fastas = {'pFC53':'pFC53.fa', 'pFC8':'pFC8.fa', 'pFC19FIXED':'pFC8.fa'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd4a4a16-1118-4fc1-a4c0-a2adc9c3442c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#List all BED files\n",
    "bed_files = [f for f in os.listdir(folder_path_bed) if f.endswith('.bed')]\n",
    "\n",
    "#Create libraries for the Bed Files\n",
    "Libraries=[]\n",
    "\n",
    "for file in bed_files:\n",
    "\n",
    "    #Read BED file\n",
    "    bed_df = pd.read_csv(os.path.join(folder_path_bed, file), sep=\"\\t\")\n",
    "\n",
    "    #Rename BED file columns\n",
    "    bed_df.columns = [\"chrom\", \"start\", \"end\", \"extra_info\", \"extra_bit\", \"direction\"]\n",
    "    \n",
    "    #Build sequence table around R-loop starts\n",
    "    Table = produce_table(bed_df, folder_path_fasta, dictionary_fastas,max_window_size=max_window_side)\n",
    "    \n",
    "    # Generate combined k-mers\n",
    "    words, freqs = get_words_combined(Table, max_w=max_window_side)\n",
    "\n",
    "    # Normalize frequencies\n",
    "    freqs=np.array(freqs)/np.max(freqs)\n",
    "    \n",
    "    # Build Library DataFrame and select top N words\n",
    "    Table_Words=pd.DataFrame({'Words':words, 'Freqs':freqs})\n",
    "    Table_Words=Table_Words.sort_values('Freqs',ascending=False).reset_index(drop=True).iloc[0:max_N_words]\n",
    "\n",
    "    # Get experiment name from file\n",
    "    exp_name = file.replace('.bed', '')\n",
    "    \n",
    "    # Append to libraries list\n",
    "    Libraries.append({'exp_name': exp_name, 'table': Table_Words})\n",
    "     \n",
    "#Save libraries\n",
    "output_path = 'data/preprocessed/R_loops_14-mers_starting_site.pickle'\n",
    "with open(output_path, 'wb') as handle:\n",
    "    pickle.dump(Libraries, handle)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
