{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8667f656-7a9a-4e28-9cf9-20bfc798a0cf",
   "metadata": {},
   "source": [
    "# 1. Preprocess Data\n",
    "\n",
    "In this notebook we preprocess genomic data associated with DNA double-strand break (DSB) and repair experiments in human and yeast cells performed by [Jeon et al.](https://www.nature.com/articles/s41467-024-51457-9).\n",
    "\n",
    "In these experiments, DSBs are introduced into DNA molecules of known sequence at distinct, prescribed positions (A, B, AB, or C/D), together with additional controlled modifiers. After break induction and repair, the resulting DNA sequences are collected via high-throughput sequencing.\n",
    "\n",
    "Each experiment is defined by a combination of the following parameters:\n",
    "\n",
    "- **Break location**: DNA cuts are introduced at specific CRISPR target sites.  \n",
    "  These are labeled **A** or **B** (single cuts), **AB** (two cuts forming a gap), or **C/D** (alternative cut sites in a different construct).\n",
    "\n",
    "- **RNA orientation**: The RNA transcript associated with the DNA can be  \n",
    "  **sense** (same orientation as the DNA sequence) or **antisense** (complementary orientation).\n",
    "\n",
    "- **Cellular background**:  \n",
    "  **WT (wild type)** cells have intact repair machinery, while **KO (knock-out)** cells lack a specific RNA/DNA-processing enzyme (RNase H), altering repair dynamics.\n",
    "\n",
    "- **RNase condition (R1 vs R2)**:  \n",
    "  These labels distinguish different RNase H contexts (e.g., normal expression versus altered or deleted activity), allowing the role of RNA–DNA interactions in repair to be tested.\n",
    "\n",
    "- **Branch**:  \n",
    "  Indicates whether the RNA transcript contains an intron branch point.  \n",
    "  **BranchΔ** denotes deletion of the branch site, preventing RNA splicing and modifying RNA structure.\n",
    "\n",
    "- **CMV**:  \n",
    "  Refers to the presence of a strong **CMV promoter**, which increases transcription levels.  \n",
    "  Removing CMV lowers RNA production while keeping the DNA sequence unchanged.\n",
    "\n",
    "For each combination of parameters, the experiment is performed with **four biological replicates**. \n",
    "\n",
    "After sequencing, the repaired DNA sequences are aligned to the original reference sequence. Reads with too many substitutions are discarded as noise. The raw data used here (stored in `data/raw`) was generated using the pipeline of [Channagiri et al.](https://github.com/tchannagiri/DSBplot), producing files named `window_freq_withoutSubst.tsv`. These files can be reproduced from the original sequencing data by running the code in  \n",
    "[Jeon et.al](https://github.com/xph9876/RNA-mediated_DSB_repair/) (see the folder `NHEJ/data_3_window`).\n",
    "\n",
    "We process this data as follows:\n",
    "1. For each table, we isolate the DNA subsequences inserted at the repair site.\n",
    "2. For individual replicates, we discard sequences with frequency below $10^{-5}$.\n",
    "3. To compute an average over the four replicates for each experimental design, we:\n",
    "   - retain only sequences appearing in all four replicates, and\n",
    "   - retain only sequences with **average frequency** of at least $10^{-5}$.\n",
    "\n",
    "The resulting processed datasets are saved as `.pickle` files under `data/preprocessed` for efficient use in the next steps.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bf6681d2-4724-4a0a-b6c7-aa8f930e8ccc",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle as pkl\n",
    "\n",
    "from utils.preprocess import format_experiment_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3a93124d-7a94-4d9c-952a-a68683f6e950",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "file_root='data/raw/'\n",
    "\n",
    "#Folders with the data\n",
    "Human_Files=[\"KO_sgAB_R1_branch\",\n",
    "\"KO_sgAB_R1_cmv\",\n",
    "\"KO_sgAB_R1_sense\",\n",
    "\n",
    "\"KO_sgAB_R2_branch\",\n",
    "\"KO_sgAB_R2_cmv\",\n",
    "\"KO_sgAB_R2_sense\",\n",
    "\n",
    "\"KO_sgA_R1_branch\",\n",
    "\"KO_sgA_R1_cmv\",\n",
    "\"KO_sgA_R1_sense\",\n",
    "\n",
    "\"KO_sgB_R2_branch\",\n",
    "\"KO_sgB_R2_cmv\",\n",
    "\"KO_sgB_R2_sense\",\n",
    "\n",
    "\"WT_sgAB_R1_branch\",\n",
    "\"WT_sgAB_R1_cmv\",\n",
    "\"WT_sgAB_R1_sense\",\n",
    "\n",
    "\"WT_sgAB_R2_branch\",\n",
    "\"WT_sgAB_R2_cmv\",\n",
    "\"WT_sgAB_R2_sense\",\n",
    "\n",
    "\"WT_sgA_R1_branch\",\n",
    "\"WT_sgA_R1_cmv\",\n",
    "\"WT_sgA_R1_sense\",\n",
    "\n",
    "\"WT_sgB_R2_branch\",\n",
    "\"WT_sgB_R2_cmv\",\n",
    "\"WT_sgB_R2_sense\",\n",
    "\n",
    "\"WT_sgCD_R1_antisense_merged\",\n",
    "\"WT_sgCD_R1_splicing_merged\",\n",
    "\n",
    "\"WT_sgCD_R2_antisense_merged\",\n",
    "\"WT_sgCD_R2_splicing_merged\"]\n",
    "\n",
    "Yeast_Files=[ \"YRNS3_YsgCD_R1_antisense\",\n",
    "\"YRNS3_YsgCD_R1_branch\",\n",
    "\n",
    "\"YRNS3_YsgCD_R2_antisense\",\n",
    "\"YRNS3_YsgCD_R2_branch\",\n",
    "\n",
    "\"YWTS3_YsgCD_R1_antisense\",\n",
    "\"YWTS3_YsgCD_R1_branch\",\n",
    "\n",
    "\"YWTS3_YsgCD_R2_antisense\",\n",
    "\"YWTS3_YsgCD_R2_branch\",\n",
    "\n",
    "\"YWT_YsgCD_R1_antisense\",\n",
    "\"YWT_YsgCD_R1_branch\",\n",
    "\n",
    "\"YWT_YsgCD_R2_antisense\",\n",
    "\"YWT_YsgCD_R2_branch\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "73eb42b2-1007-4bae-92e4-a8656a3cac2b",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def extract_tables(Files, min_threshold):\n",
    "    experiments_all=[]\n",
    "    experiments_average=[]\n",
    "\n",
    "    exp_id=0\n",
    "    for i in range(len(Files)):\n",
    "        experimental_design_name=Files[i] \n",
    "        file_name=experimental_design_name+\"/window_freq_withoutSubst.tsv\"\n",
    "        \n",
    "        #reformat name\n",
    "        experimental_design_name='_'.join(experimental_design_name.split('_')[:4])\n",
    "        exp_id+=1\n",
    "        \n",
    "        Experiment_Table=format_experiment_table(file_root, file_name)\n",
    "        Libraries=Experiment_Table.library.unique()\n",
    "        \n",
    "        #Individual Libraries\n",
    "        for lib in Libraries:\n",
    "            Replicate=Experiment_Table[Experiment_Table['library'] == lib].drop(columns='library')\n",
    "            \n",
    "            #Use Threshold\n",
    "            Replicate=Replicate[Replicate['freq']>=min_threshold]\n",
    "                    \n",
    "        \n",
    "            #Save replicate\n",
    "            replicate_name=experimental_design_name+'_'+lib\n",
    "            experiments_all.append({'exp_name':replicate_name, 'table':Replicate.copy()})\n",
    "        \n",
    "        #Combined Table\n",
    "        Combined = Experiment_Table.\\\n",
    "            pivot_table(index='sequence', columns='library', values='freq', fill_value=0).\\\n",
    "                reset_index()\n",
    "        \n",
    "        # Compute min and average across the library columns\n",
    "        library_cols = Combined.columns.drop('sequence')  # all library columns\n",
    "        Combined['min_freq'] = Combined[library_cols].min(axis=1)\n",
    "        Combined['avg_freq'] = Combined[library_cols].mean(axis=1)\n",
    "        \n",
    "        #Apply Threshold\n",
    "        #Only consider cases where the word appears in four replicates\n",
    "        Combined=Combined[Combined['min_freq']>0]\n",
    "        Combined=Combined[Combined['avg_freq']>=min_threshold]\n",
    "        \n",
    "        #Save table\n",
    "        experiments_average.append({'exp_name':experimental_design_name, 'table':Combined.copy()})\n",
    "\n",
    "    return experiments_all, experiments_average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5eec7765-6252-43aa-a11b-922fbed577a5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "#Format Human Files\n",
    "min_threshold=0\n",
    "Human_experiments_all,Human_experiments_average=extract_tables(Human_Files, min_threshold)\n",
    "\n",
    "#Save Tables with pickle objects\n",
    "with open('data/preprocessed/Human_experiments_all.pickle', 'wb') as handle:\n",
    "    pkl.dump(Human_experiments_all, handle)\n",
    "with open('data/preprocessed/Human_experiments_average.pickle', 'wb') as handle:\n",
    "    pkl.dump(Human_experiments_average, handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d86a1647-e5dd-40cd-a2a1-70ffde860ccf",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "#Format Yeast Files\n",
    "min_threshold=0\n",
    "Yeast_experiments_all,Yeast_experiments_average=extract_tables(Yeast_Files, min_threshold)\n",
    "\n",
    "#Save Tables with pickle objects\n",
    "with open('data/preprocessed/Yeast_experiments_all.pickle', 'wb') as handle:\n",
    "    pkl.dump(Yeast_experiments_all, handle)\n",
    "with open('data/preprocessed/Yeast_experiments_average.pickle', 'wb') as handle:\n",
    "    pkl.dump(Yeast_experiments_average, handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1891b80-c938-42b0-9c62-b3b0be1df3a9",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
